File: array/problem153/problem153.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int findMin(vector<int>& nums) {
        // Binary search using right boundary comparison
        int lo = 0, hi = static_cast<int>(nums.size()) - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] > nums[hi]) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return nums[lo];
    }
};

File: array/problem153/problem153.py
from typing import List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        # Binary search: compare mid with right to locate the rotated minimum
        lo, hi = 0, len(nums) - 1
        while lo < hi:
            mid = (lo + hi) // 2
            if nums[mid] > nums[hi]:
                # Minimum is in (mid, hi]
                lo = mid + 1
            else:
                # Minimum is in [lo, mid]
                hi = mid
        return nums[lo]
